server/ComputeSuperpixelFeatures/ComputeSuperpixelFeatures.py

```diff
@@ -0,0 +1,387 @@
+from cli_common import utils as cli_utils  # noqa
+
+
+def compute_input_tensors(slide_path, tile_position, args, **it_kwargs):
+
+    # get slide tile source
+    ts = large_image.getTileSource(slide_path)
+
+    # get current magnification
+    magnification = ts.getMetadata()['magnification']
+
+    # get requested tile
+    tile_info = ts.getSingleTile(
+        tile_position=tile_position,
+        format=large_image.tilesource.TILE_FORMAT_NUMPY,
+        **it_kwargs)
+
+    # get tile image
+    im_tile = tile_info['tile'][:, :, :3]
+
+    # get current left and top positions
+    tile_left_position = tile_info['gx']
+    tile_top_position = tile_info['gy']
+
+    # perform color normalization
+    im_nmzd = htk_cnorm.reinhard(im_tile,
+                                 args.reference_mu_lab,
+                                 args.reference_std_lab)
+
+    # perform color decovolution
+    w = cli_utils.get_stain_matrix(args)
+
+    im_stains = htk_cdeconv.color_deconvolution(im_nmzd, w).Stains
+
+    im_hematoxylin_stain = im_stains[:, :, 0].astype(np.float)/255.0
+    im_eosin_stain = im_stains[:, :, 1].astype(np.float)/255.0
+
+    # get the number of super-pixels
+    im_width, im_height = im_tile.shape[:2]
+    n_superpixels = (im_width/args.patchSize) * (im_height/args.patchSize)
+
+    # perform SLIC to get labels with max compactness
+    # slic has a value 0 as a first label, so added 1
+    im_label = slic(im_nmzd, n_segments=n_superpixels, compactness=100) + 1
+
+    # extract features
+    region_props = regionprops(im_label)
+
+    # numFeatures from hematoxylin and eosin stained images
+    n_labels = len(region_props)
+
+    # we will use feature shape: NHWC on CPU, NCHW on GPU
+    # with 2 channels: hematoxylin adn eosin
+    fdata = np.zeros((n_labels, args.patchSize, args.patchSize, 2))
+
+    # set an array for x, y centers
+    cdata = np.zeros((n_labels, 2))
+
+    for i in range(n_labels):
+        cen_x, cen_y = region_props[i].centroid
+
+        # get bounds of region
+        min_row, max_row, min_col, max_col = \
+            get_patch_bounds(cen_x, cen_y, args.patchSize, im_width, im_height)
+
+        # get image centers at low-res
+        cdata[i, 0] = cen_x * (magnification / args.analysis_mag) + \
+            tile_left_position
+        cdata[i, 1] = cen_y * (magnification / args.analysis_mag) + \
+            tile_top_position
+
+        # get image patches
+        fdata[i, :, :, 0] = im_hematoxylin_stain[min_row:max_row, min_col:max_col]
+        fdata[i, :, :, 1] = im_eosin_stain[min_row:max_row, min_col:max_col]
+
+    return cdata, fdata
+
+
+def get_encoded_features(input_tensors, input_model, args):
+
+    # load input model
+    autoencoder = load_model(input_model)
+
+    # input: the first input of the model, output: the last output of the model
+    encoder = Model(inputs=autoencoder.layers[0].input, outputs=autoencoder.layers[6].output)
```

Comment:
Layer indexes are hard-coded

Reply:
Added some lines to detect encoding layers